iguana.corecontroller/src/main/java/org/aksw/iguana/cc/utils/IndexedLineReader.java

```diff
@@ -0,0 +1,204 @@
+     * If the separator is blank, every bundle between two blank lines will be indexed. <br/>
+     * If the separator parameter is null, it will instead just index every non-blank line of the file.
+     * @param filepath path to the file
+     * @param separator the separator for each bundle
+     * @throws IOException
+     */
+    private IndexedLineReader(String filepath, String separator) throws IOException {
+        this.file = new File(filepath);
+
+        if(separator == null) {
+            this.indexFile();
+        }
+        else {
+            this.indexFile(separator);
+        }
+    }
+
+    /**
+     * If a separator wasn't given, this method returns the line with the corresponding index inside the file. <br/>
+     * If a separator was given, this method returns a string of multiple lines, that are between two separators
+     * (including the beginning and end of file), with the corresponding index.
+     * @param index the index of the searched line or bundle of lines
+     * @return the searched line or bundle of lines
+     * @throws IOException
+     */
+    public String readLine(int index) throws IOException {
+        // conversion from long to int (lines shouldn't be bigger than ~2GB)
+        byte[] data = new byte[Math.toIntExact(this.indices.get(index)[1])];
+        String output;
+        try(RandomAccessFile raf = new RandomAccessFile(this.file, "r")) {
+            raf.seek(this.indices.get(index)[0]);
+            raf.read(data);
+            output = new String(data, StandardCharsets.UTF_8);
+        }
+        return output;
+    }
+
+    /**
+     * This method returns a list of strings that contains every indexed line or bundle of lines.
+     * @return list of lines
+     * @throws IOException
+     */
+    public List<String> readLines() throws IOException {
+        ArrayList<String> out = new ArrayList<>();
+        for(int i = 0; i < indices.size(); i++) {
+            out.add(this.readLine(i));
+        }
+        return out;
+    }
+
+    /**
+     * Returns the number of indexed non-blank lines or bundle of lines (depends on if a separator was given).
+     * @return number of indexed objects
+     */
+    public int size() {
+        return this.size;
+    }
+
+    /**
+     * Indexes every non-blank line inside the file.
+     * @throws IOException
+     */
+    private void indexFile() throws IOException {
+        this.indices = new ArrayList<>();
+        try(FileReader fr = new FileReader(this.file, StandardCharsets.UTF_8)) {
+            BufferedReader br = new BufferedReader(fr);
+            // The method needs to know the length of the line ending used in the file to be able to properly calculate
+            // the starting byte position of a line
+            int lineEndingLength = FileUtils.getLineEnding(this.file.getAbsolutePath()).length();
+            long index = 0;
+            String line;
+            while((line = br.readLine()) != null) {
```

Comment:
this should also be read character-wise and not line-wise. Same implementation as in indexFile(String separator) can be used.